### ISSUE TITLE
Build PoC for streamfields-are-rows with SASS width rules

### ISSUE THREAD

**Issue by Pomax** [mozilla/foundation.mozilla.org#6595]
See  https://github.com/mozilla/foundation.mozilla.org/issues/6517 for the SASS details and code

**Comment by Pomax** [comment]
Some investigations:

![image](https://user-images.githubusercontent.com/177243/118870464-70645f80-b89b-11eb-97ad-d4f588c23cbc.png)

Testing with primary page:

```jinja
...

{% block content %}
  {% include "partials/multipage_nav_mobile.html" %}

  {% primary_page_menu page %}

  <div class="cms">
    {% for block in page.body %}
    <div class="container {{block.block_type}}-block">
      <div class="row">
        <div class="{% if page.specific.narrowed_page_content %}col-lg-8 offset-lg-2{% else %}col-12{% endif %}">
          {% if block.block_type == 'heading' %}
            <h1>{{ block.value }}</h1>
          {% else %}
            {% include_block block %}
          {% endif %}
        </div>
      </div>
    </div>
    {% endfor %}
  </div>
{% endblock %}
```

seems to work well enough. To escape to 100% page width content, we would need to be able to toggle to:

```
    <div class="container {{block.block_type}}-block m0 p0 w-100 mw-100">
      <div class="row justify-content-center no-gutters">
        <div class="w-100">
        ...
        </div>
      </div>
    </div>
```

where the `no-gutters` class on the row is rather important, and the content is not sized to "a column count" but to "100% width" instead, with the container sized to "width: 100%" _and_ "max-width: 100%". Without the mw-100, this will not work.

**Comment by danielfmiranda** [comment]
Hi Everyone, here are my findings from trying to use the suggested layout from the comments above:

One of the main templates that I tried to work with was the primary_page.html, and was able to get it very close to the original, for a comparison visit: [ Percy Comparison](https://percy.io/Mozilla-Foundation/foundation.mozilla.org/builds/10848455/changed/610060082?browser=chrome&browser_ids=14%2C16&subcategories=unreviewed%2Cchanges_requested&viewLayout=side-by-side&viewMode=new&width=1280&widths=375%2C1280 )

The only difference between the two pages would be the spacing on the new version, since the stream blocks are wrapped in containers we would need to escape the default margin/padding that bootstrap containers use. 

I then attempted this on the blog page template, and again the results were [pretty close](https://percy.io/Mozilla-Foundation/foundation.mozilla.org/builds/10848455/changed/610060250?browser=chrome&browser_ids=14%2C16&subcategories=unreviewed%2Cchanges_requested&viewLayout=side-by-side&viewMode=new&width=1280&widths=375%2C1280), one thing I did want to note however is that some of the templates that the blog page extends wrap the page in one large containing div, so when it rendered the stream_blocks in their own containers, it added even more padding and making them smaller than the rest of the content. I also found that some templates share these extended pages so updating them to work with the new streamblock layout could be tricky.

My initial thinking was that this could be worked around by removing margins and padding for the stream block containers, however this would then reintroduce the issue of not being able to have a "End to end" wide image block since it would be constrained in the outer most container. 

I was able to resolve this by wrapping the share buttons and heading in its own container div, closing it, and then allowing the for loop to render all the blocks in their own containers. Followed by reopening a new container for things such as the share buttons and suggested blog posts at the bottom of the page. 

One thing I did like about this format though was that I was able to wrap the for loop rendering all the stream blocks in a containing div named "blog-body": 

```
<div class="blog-body">
    {% for block in page.body %}
    {% if block.block_type == 'paragraph'%}
 <!-- Rendering paragraph blocks since they dont have their own template file --> 
    <div class="container {{block.block_type}}-block">
      <div class="row justify-content-center">
        <div class="streamfield-content">
          {% if block.block_type == 'heading' %}
            <h1>{{ block.value }}</h1>
          {% else %}
            {% include_block block with parent_page=page page_type="blog"%}
          {% endif %}
        </div>
      </div>
    </div>
  {% else %}
 <!-- Rendering all other blocks in their respective template files--> 
    {% include_block block with parent_page=page page_type="blog" %}
  {% endif %}    
  {% endfor %}
  </div>
```

And give all the stream blocks a class of streamblock-content like so: 

```
<div class="container {{block.block_type}}-block">
  <div class="row justify-content-center">
    <div class="streamfield-content">
...
    </div>
  </div>
</div>
```

Which means I was able to easily access and change their widths using css like so: 
```
.blog-body {
  .streamfield-content {
    @extend .col-8;
  }
}
```

This seemed nice to me as if we ever decided to make a blog template wider, we can just visit the css file and change it to a col-10 or 12, and all of our block items would follow accordingly. 

Or, if we wanted to implement the ability to let the user choose between wide and narrow content layout, we can just add an additional if else in the template to add classes, and then choose the widths in css like so, there is a rough example of doing this in the primay page template in the PR #6796:

```
.primary-page-content {
  &.wide {
    .streamfield-content {
      @extend .col-12;
    }
  }

  &.narrow {
    .streamfield-content {
      @extend .col-8;
    }
  }
}
```


So to sum up I think this streamfields-as-rows would be a nice way to go and easy to maintain once we update the templates accordingly, however, it does seem like updating some templates such as campaigns might be a little tricky. 

If you would like to see what I did in the proof of concept or view some of the pages using this new format in the review app, I have a connected pr #6796, however I did schedule a meeting with @benhohner @KalobTaulien and @Pomax to go over my findings, explain what I did in the PR, and discuss next steps this coming monday the 7th!

**Comment by danielfmiranda** [comment]
Closing as we have created [#6838](https://github.com/mozilla/foundation.mozilla.org/issues/6838)
